src/MatrixProductOperator.jl

```diff
@@ -0,0 +1,193 @@
+    elseif (order[:l] < order[:o] < order[:i])
+        (vinds[(n - 1, n)], oinds[n], iinds[n])
+    elseif order[:i] < order[:l] < order[:o]
+        (iinds[n], vinds[(n - 1, n)], oinds[n])
+    elseif order[:o] < order[:l] < order[:i]
+        (oinds[n], vinds[(n - 1, n)], iinds[n])
+    elseif order[:i] < order[:o] < order[:l]
+        (iinds[n], oinds[n], vinds[(n - 1, n)])
+    else
+        (oinds[n], iinds[n], vinds[(n - 1, n)])
+    end
+    push!(tn, Tensor(last(arrays), labels))
+
+    # add other tensors
+    for (i, data) in zip(2:n-1, arrays[2:end-1])
+        lind = vinds[(mod1(i - 1, n), i)]
+        rind = vinds[(i, mod1(i + 1, n))]
+
+        labels = [lind, rind, iinds[i], oinds[i]]
+        permute!(labels, permutator)
+
+        tensor = Tensor(data, labels)
+        push!(tn, tensor)
+    end
+
+    # mark input indices
+    for (site, label) in iinds
+        index = inds(tn, label)
+        index.meta[:site] = site
+        index.meta[:plug] = :input
+    end
+    # mark output indices
+    for (site, label) in oinds
+        index = inds(tn, label)
+        index.meta[:site] = site
+        index.meta[:plug] = :output
+    end
+
+    println("$tn.tensors")
+    return tn
+end
+
+function MatrixProductOperator{Closed}(arrays; χ = nothing, order = (:l, :r, :i, :o), meta...)
+    !issetequal(order, (:l, :r, :i, :o)) && throw(ArgumentError("`order` must be a permutation of the :l, :r, :i and :o"))
+    order = Dict(side => i for (i, side) in enumerate(order))
+
+    # check format
+    # TODO check χ?
+    if !all(==(4) ∘ ndims, arrays)
+        throw(DimensionMismatch("virtual bond-dims mismatch"))
+    end
+
+    tn = TensorNetwork{MatrixProductOperator{Closed}}(; χ = χ, order = order, meta...)
+
+    n = length(arrays)
+    vinds = Dict(x => Symbol(uuid4()) for x in ringpeek(1:n))
+    oinds = Dict(i => Symbol(uuid4()) for i in 1:n)
+    iinds = Dict(i => Symbol(uuid4()) for i in 1:n)
+    permutator = [order[i] for i in (:l, :r, :i, :o)]
+
+    # add tensors
+    for (i, data) in enumerate(arrays)
+        lind = vinds[(mod1(i - 1, n), i)]
+        rind = vinds[(i, mod1(i + 1, n))]
+
+        labels = [lind, rind, iinds[i], oinds[i]]
+        permute!(labels, permutator)
+
+        tensor = Tensor(data, labels)
+        push!(tn, tensor)
+    end
+
+    # mark input indices
+    for (site, label) in iinds
+        index = inds(tn, label)
+        index.meta[:site] = site
+        index.meta[:plug] = :input
+    end
+    # mark output indices
+    for (site, label) in oinds
+        index = inds(tn, label)
+        index.meta[:site] = site
+        index.meta[:plug] = :output
+    end
+
+    return tn
+end
+ 
+
+struct MPOSampler{B<:Bounds,T} <: Random.Sampler{TensorNetwork{MatrixProductOperator{B}}}
+    n::Int
+    ip::Int
+    op::Int
+    χ::Int
+end
+
+Base.eltype(::MPOSampler{B}) where {B<:Bounds} = TensorNetwork{MatrixProductOperator{B}}
+
+function Base.rand(
+    ::Type{MatrixProductState{B}},
```

Comment:
I think that you forgot to add this suggestion to the batch

Reply:
Completely. Thanks!
```suggestion
    ::Type{MatrixProductOperator{B}},
```